pkg/api/pod/util.go

```diff
@@ -1284,7 +1284,7 @@ func MarkPodProposedForResize(oldPod, newPod *api.Pod) {
 		if c.Resources.Requests == nil {
 			continue
 		}
-		if cmp.Equal(oldPod.Spec.Containers[i].Resources, c.Resources) {
+		if reflect.DeepEqual(oldPod.Spec.Containers[i].Resources, c.Resources) {
```

Comment:
Do we need to equate `nil` with empty for maps?  I don't think that is handled by reflect.

Perhaps we make a new function, e.g. `EqualResources(lhs, rhs)` which encapsulates this logic, has a great big comment about why it is important, and has a robust unit test?

Reply:
How about this?

```
apiequality.Semantic.DeepEqual(oldPod.Spec.Containers[i].Resources, c.Resources)
```

Learnt from https://github.com/carlory/kubernetes/blob/master/pkg/controller/resourcequota/resource_quota_controller.go#L232

Reply:
This says it panics when comparing unexported fields - is this an issue?

Reply:
So does go-cmp if you don't take care, doesn't it?

Reply:
Yes, which is what I wanted to remedy with this PR. Perhaps the best option is to normalise the structs before comparison (that is, explicitly set all empty maps to nil, or vice versa), and then just use DeepEqual?

Reply:
Going with a normalizaton + `reflect.DeepEqual` in a central method sounds feasible. It does require reflection though.

Reply:
Semantic.DeepEqual says:

```
// Unexported field members cannot be compared and will cause an informative panic; you must add an Equality                                      
// function for these types.
```

If that is the *only* panic case, we should be able to add tests that prove that Pods are safe, right?  Then, if anyone does something in the future that is unsafe, the tests will fail.

e.g. similar to the round-tripper, do 1000 (or more if we need) loops of:
* fuzz a pod "lhs"
* fuzz a pod "rhs"
* set their names to different values (just in case fuzzer made exactly the same pod)
* run Semantic.DeepEqual and assert that it must return false

If it panics, we know we have a problem.  Repeat for any type we need DeepEqual.  Or even better, do it for all registered types via `scheme` ?

Reply:
I do not want to have to write "normalization" functions.  The alternative is to code-generate DeepEqual methods like we do for DeepCopy.  That's not SO bad, but is a much larger effort.

Reply:
I talked to Joe briefly, and he agreed:  First let's see if we can convince ourselves that Semantic DeepEqual is safe for all API types.  If that is impossible or not performant enough (data!) then we could look at code-gen.